Updated Documentation For Change in Behaviour of Reading X-Ray Tracing Info From Lambda

jdoherty opened the pull request:
Related issues #

https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8368

Related [OTEP(s)](https://github.com/open-telemetry/oteps) #

@jsuereth asked I open this change here - rather than over in the documentation repo (https://github.com/open-telemetry/opentelemetry-specification/pull/3502) 

@tylerbenson , @carlosalberto, @Oberon00 FYI I know you previously reviewed the other PR so this is now open here as requested by Josh

jdoherty commented on specification/trace/semantic_conventions/instrumentation/aws-lambda.md:
```diff
@@ -56,13 +56,15 @@ and the [cloud resource conventions][cloud]. The following AWS Lambda-specific a
-contain an incomplete trace context which indicates X-Ray isn’t enabled. The environment variable will be set and the
+There are two common ways to read the tracing information if AWS X-Ray is enabled for a lambda function:
+
+1. Via the system property `com.amazonaws.xray.traceHeader` in the event of a Java function.
```
Hi Tyler, is this any better? If not, could you please offer a suggested alternative? I'm trying to help the community but seem to be getting the run around somewhat on this.

jsuereth reviewed (approved):
LGTM pending @tylerbenson approval as well.

tylerbenson commented on specification/trace/semantic_conventions/instrumentation/aws-lambda.md:
```diff
@@ -66,6 +66,9 @@ contain an incomplete trace context which indicates X-Ray isn’t enabled. The e
 `Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can
 disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
 
+**Note**: If you are instrumenting a Java AWS lambda then instrumentation will try to parse an OpenTelemetry `Context` out of the system property `com.amazonaws.xray.traceHeader` using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) first before checking and attempting to parse the environment variable ab ...
```
```suggestion
**Note**: When instrumenting a Java AWS Lambda then instrumentation SHOULD first try to parse an OpenTelemetry `Context` out of the system property `com.amazonaws.xray.traceHeader` using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) before checking and attempting to parse the environment variable above.
```

tylerbenson reviewed (approved):
Minor wording suggestions but otherwise looks good to me.

jdoherty commented on specification/trace/semantic_conventions/instrumentation/aws-lambda.md:
```diff
@@ -56,18 +56,19 @@ and the [cloud resource conventions][cloud]. The following AWS Lambda-specific a
 `Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can
 disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
 
+**Note**: If you are instrumenting a Java AWS lambda then instrumentation will try to parse an OpenTelemetry `Context` out of the system property `com.amazonaws.xray.traceHeader` using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) first before checking and attempting to parse the environment variable ab ...
```
Hi Christian, there is no advantage.  This is an alternative way to achieve the same thing.  This was opened to address this: 

https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/7579

Long story short - you need hacky solutions to set the environment variable and its now highly discouraged by Oracle in newer versions of Java.

smirnoal replied:
Lambda Java 17 Runtime and further will only use system properties approach. That's the main advantage, to make OTEL work with Lambda properly

smirnoal commented:
> Is there any AWS Lambda documentation that mentions the use (and name) of the system property?

not yet, but I'm working with AWS docs team on this to be documented. 
As for now it is used in the source codes for Lambda Java Runtime and AWS SDK (both open source)

trask commented:
> As for now it is used in the source codes for Lambda Java Runtime and AWS SDK (both open source)

can you add links to these usages? that's good enough for me, thx

smirnoal commented:
> > As for now it is used in the source codes for Lambda Java Runtime and AWS SDK (both open source)
> 
> can you add links to these usages? that's good enough for me, thx

here you are, Trask:

AWS x-ray SDK:
https://github.com/aws/aws-xray-sdk-java/blob/711e48cfd54400e728f99bec3bb3baf417276b46/aws-xray-recorder-sdk-core/src/main/java/com/amazonaws/xray/contexts/LambdaSegmentContext.java#L40


AWS Lambda Java runtime:
https://github.com/aws/aws-lambda-java-libs/blob/main/aws-lambda-java-runtime-interface-client/src/main/java/com/amazonaws/services/lambda/runtime/api/client/AWSLambda.java#L62

jdoherty commented:
Hi @jsuereth now we have merged https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8368
and everyone on this chain has approved this one - can you ask a fellow approver / maintainer to take a look please?

jsuereth commented:
@jdoherty Looks good, can you add a changelog?

jdoherty commented:
> Can you fix the lint errors? I'll merge as soon as I see green.

Thank you :-).  Darn whitespace.  I've now removed this and rebased the other changes from mainline.  Running the lint locally shows no errors for me now.  

Do you need to kick off the workflow? I see my push does not trigger it.

Oberon00 commented on specification/trace/semantic_conventions/instrumentation/aws-lambda.md:
```diff
@@ -66,6 +66,9 @@ contain an incomplete trace context which indicates X-Ray isn’t enabled. The e
 `Context` will be valid and sampled only if AWS X-Ray has been enabled for the Lambda function. A user can
 disable AWS X-Ray for the function if the X-Ray Span Link is not desired.
 
+**Note**: When instrumenting a Java AWS Lambda, instrumentation SHOULD first try to parse an OpenTelemetry `Context` out of the system property `com.amazonaws.xray.traceHeader` using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) before checking and attempting to parse the environment variable above.
+As above, if the resulting `Context` is valid then a Span Link SHOULD be added to the new Spans start options.
```
```suggestion
**Note**: When instrumenting a Java AWS Lambda, instrumentation SHOULD first try to parse an OpenTelemetry `Context` out of the system property `com.amazonaws.xray.traceHeader` using the [AWS X-Ray Propagator](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md) before checking and attempting to parse the environment variable above.
```

smirnoal commented:
> > Is there any AWS Lambda documentation that mentions the use (and name) of the system property?
> 
> not yet, but I'm working with AWS docs team on this to be documented. As for now it is used in the source codes for Lambda Java Runtime and AWS SDK (both open source)

docs were updated: https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime
